### ISSUE TITLE
Do you have an example with custom types extending ObjectType?

### ISSUE THREAD

**Issue by tomasvts** [thecodingmachine/graphqlite#277]
My situation is the following:
I already have a graphql codebase made with webonyx/graphql-php and Yii2 framework with lots of custom types extending ObjectType, and I was thinking of migrating to grapqhlite using what I'm reading at https://graphqlite.thecodingmachine.io/docs/custom-types but with no avail. 

I'm looking for something like what you have ScalarTypes: with https://github.com/thecodingmachine/graphqlite-misc-types/blob/master/src/AnyScalar/AnyScalarType.php but with ObjectType instead of ScalarType types.

**Comment by moufmouf** [comment]
Hey @tomasvts ,

I just had a look at the documentation and it seems it needs to be improved a bit.

I'm assuming you are bootstrapping GraphQLite using the `SchemaFactory` class, right?

Could you try this?

```php
$staticTypeMapper = new StaticTypeMapper();

// Let's register a type that maps by default to the "MyClass" PHP class
$staticTypeMapper->setTypes([
    MyClass::class => new MyCustomOutputType()
]);

// If you don't want your output type to map to any PHP class by default, use:
$staticTypeMapper->setNotMappedTypes([
    new MyCustomOutputType()
]);

$schemaFactory->addTypeMapper($staticTypeMapper);
```

(the last line is missing in the documentation)

Could you let me know if this works?

**Comment by tomasvts** [comment]
There is some progress, yes!

But something is wrong with my ObjectType implementation because I'm now seeing this failure

```
Return value of TheCodingMachine\\GraphQLite\\Mappers\\StaticTypeMapper::mapClassToType() 
must implement interface TheCodingMachine\\GraphQLite\\Types\\MutableInterface, 
instance of App\\Infrastructure\\APISchema\\GraphQL\\Type\\ShowStatusType returned',
```

My ShowStatusType extends webonyx/graphql-php ObjectType but it does not implement MutableInterface

```
<?php declare(strict_types=1);

namespace App\Infrastructure\APISchema\GraphQL\Type;

use DateHelper;
use GraphQL\Type\Definition\ObjectType;
use GraphQL\Type\Definition\ResolveInfo;
use App\Domain\Show\Show;
use App\Infrastructure\APISchema\GraphQL\Type\Enum\ShowConnectionEnum;
use App\Infrastructure\APISchema\GraphQL\Type\Enum\ShowStatusEnum;
use App\Infrastructure\APISchema\GraphQL\Types;

class ShowStatusType extends ObjectType
{
```

If I make it implement MutableInterface it fails with:

```
PHP Fatal error:  Declaration of GraphQL\Type\Definition\ObjectType::getFields() must be compatible with TheCodingMachine\GraphQLite\Types\MutableInterface::getFields(): array in /App/src/Infrastructure/APISchema/GraphQL/Type/ShowStatusType.php on line 14
```

**Comment by moufmouf** [comment]
Oh! I see!

Indeed, I refactored the `StaticTypeMapper` class a few months ago and failed to notice that now, it accepts only output types that are implementing `MutableInterface`. Which makes no sense in the case where you are migrating from Webonyx types.

The `MutableInterface` is important, this is what allows us to use the `@ExtendType` annotation that can be used to add fields in an existing type.

Ok, so we can sort this problem by creating an "adapter" that takes an OutputType and implements both OutputType&MutableInterface.

I won't have time to work on this before next week at best. If you want to give it a shot, I can certainly help you. Otherwise, I'll try having a look at it as soon as I can.

TODO list:

- [ ] Create an adapter that takes an OutputType and implements both OutputType&MutableInterface.
- [ ] Create an adapter that takes an InputType and implements both InputType&ResolvableMutableInputInterface
- [ ] Use those adapters automatically in `StaticTypeMapper` (these are implementation details that should be hidden to the end user)
- [ ] Fix documentation (talk about how to use this with the `SchemaFactory`)

**Comment by tomasvts** [comment]
I'm very grateful for your help and time @moufmouf  !

As I'm a little bit late with my current task and I can't estimate how long it could take me to do those adapters correctly, what I'll do now is what I know best and migrate my ObjectTypes to graphqlite compatible @Type entities, 
I don't think it will be too hard.

You are free to close now. 

Thanks

**Comment by moufmouf** [comment]
:+1: I'll keep this issue open until I find the time to create the adapter and fix the doc.
Thanks for your feedback!

**Comment by moufmouf** [comment]
In case you are still working on this, I merged a fix in the upcoming 4.1.

I'll keep this issue open because I have handled `ObjectType` and `InterfaceType` but I still need to work on `InputType`.

**Comment by oojacoboo** [comment]
@moufmouf can we wrap this up and close it as well for a 5.0 release?
